main/src/ca/uwaterloo/flix/language/phase/Kinder.scala

```diff
@@ -0,0 +1,1155 @@
+/*
+ * Copyright 2021 Matthew Lutze
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.CompilationError
+import ca.uwaterloo.flix.language.ast._
+import ca.uwaterloo.flix.language.errors.KindError
+import ca.uwaterloo.flix.util.Validation
+import ca.uwaterloo.flix.util.Validation.{ToFailure, ToSuccess, flatMapN, mapN, traverse}
+
+/**
+  * Attributes kinds to the types in the AST.
+  *
+  * For enums, classes, instances, and type aliases:
+  * Either:
+  *   - type parameters are not annotated and are then assumed all to be Star, or
+  *   - type parameters are all annotated with their kinds.
+  *
+  * For defs:
+  * Either:
+  *   - type parameters are all annotated with their kinds, or
+  *   - type parameters are not annotated and their kinds are inferred from their
+  *     use in the formal parameters, return type and effect, and type constraints.
+  *     This inference uses the following rules:
+  *       - If the type variable is the type of a formal parameter, it is ascribed kind Star.
+  *       - If the type variable is the return type of the function, it is ascribed kind Star.
+  *       - If the type variable is the effect type of the function, it is ascribed kind Bool.
+  *       - If the type variable is an argument to a type constraint, it is ascribed the class's parameter kind
+  *       - If the type variable is an argument to a type constructor, it is ascribed the type constructor's parameter kind.
+  *       - If the type variable is used as an type constructor, it is ascribed the kind Star -> Star ... -> Star -> X,
+  *         where X is the kind inferred from enacting these rules in the place of the fully-applied type.
+  *       - If there is an inconsistency among these kinds, an error is raised.
+  *
+  * In inferring types, variable type constructors are assumed to have kind * -> * -> * -> ???.
```

Comment:
I don't understand.

Reply:
Did you understand the rule:
```
  *       - If the type variable is used as an type constructor, it is ascribed the kind Star -> Star ... -> Star -> X,
  *         where X is the kind inferred from enacting these rules in the place of the fully-applied type.
```
?

This is a worse explanation of the same rule that I forgot to remove :)